[16.0][ADD] hr_timesheet_begin_end_no_end

carmenbianca opened the pull request:
Refactored some code in hr_timesheet_begin_end to be more extensible, and to be more Odoo 16-ish.

Then built a module that removes one of the constraints because of a customer request.

Depends on #692. This is based on top of that branch.

Internal task: T12630

carmenbianca committed:
[REF] hr_timesheet_begin_end: Make more extensible

The separation of checks into separate methods is needed because I want
to disable one check in another module. This makes the module more
extensible.

The unit_amount_from_start_stop method also makes the module more
extensible.

I have also moved the onchange to the top of the file, according to the
OCA contribution guidelines.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] hr_timesheet_begin_end_no_end

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] hr_timesheet_begin_end: Refactor unit_amount to be a compute field

Instead of using onchange, which is less convenient in Odoo 16.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented on hr_timesheet_begin_end/models/account_analytic_line.py:
```diff
@@ -15,16 +16,22 @@ class AccountAnalyticLine(models.Model):
 
-    @api.constrains("time_start", "time_stop", "unit_amount")
-    def _check_time_start_stop(self):
+    # Override to be a computed field.
+    unit_amount = fields.Float(
+        compute="_compute_unit_amount",
+        store=True,
+        readonly=False,
+    )
```
This reproduces as closely as possible the old behaviour. Technically `readonly=False` isn't really desirable here, I think, but it matches how it used to be when this used onchange (and there is a constraint to make sure the two stay in sync, anyway).

huguesdk replied:
as this model is also used for other things than timesheets, it is very important that `unit_amount` must not be computed if `time_start` and `time_stop` are both not set. since it is `store=True` and `readonly=False`, it should work as expected. adding a test for this would be good.

carmenbianca replied:
Did this in #692 in the `create` method. Was a bit tricky to get right.

carmenbianca commented:
@huguesdk do you want to review this?

huguesdk commented on hr_timesheet_begin_end/tests/test_timesheet_begin_end.py:
```diff
@@ -22,18 +22,20 @@ def setUp(self):
+        self.assertEqual(line.unit_amount, 4)
 
-    def test_onchange_no_update(self):
+    def test_compute_unit_amount_no_update(self):
```
this should be renamed, as it actually updates the value (sets it to 0) (even though it was already at 0).

carmenbianca replied:
Done in #692

huguesdk commented on hr_timesheet_begin_end_no_end/tests/test_timesheet_begin_end_no_end.py:
```diff
@@ -0,0 +1,49 @@
+        line = self.base_line()
+        line.update({"time_stop": 0, "unit_amount": 0})
+        # No error.
+        self.env["account.analytic.line"].create(line)
```
another test to be done: if `unit_amount` is not provided, check that it is 0.

carmenbianca replied:
Done in #692 (`test_compute_unit_amount`). Not sure if some variation of this test also needs to be here.

huguesdk commented on hr_timesheet_begin_end_no_end/tests/test_timesheet_begin_end_no_end.py:
```diff
@@ -0,0 +1,49 @@
+        line.update({"time_stop": 1, "unit_amount": 0})
+        with self.assertRaises(exceptions.ValidationError):
+            self.env["account.analytic.line"].create(line)
+
```
it would be good to add tests that (try to) change an existing line and check whether it also works correctly.

huguesdk reviewed (changes requested):
great refactor, thanks @carmenbianca!

please move the commits adding the new module to another pull request (commits on a branch on top of this one), as the refactor should be able to be accepted independently (and the version numbers correctly increased independently too).

huguesdk commented on hr_timesheet_begin_end_no_end/tests/test_timesheet_begin_end_no_end.py:
```diff
@@ -0,0 +1,54 @@
+
+    def base_line(self):
+        return {
+            "name": "test",
+            "date": fields.Date.today(),
+            "time_start": 10.0,
+            "time_stop": 12.0,
+            "user_id": self.user.id,
+            "account_id": self.analytic.id,
+            "amount": -60.0,
```
the demo data should reflect timesheet lines data, thus with a `project_id` and an `employee_id`. `user_id`, `account_id` and `amount` are not needed in this case.

carmenbianca replied:
hi @huguesdk, i am not certain i agree.

- in `hr_timesheet_begin_end`, the test data is very similar to the above. if we change it here, we should change it there.
- `project_id` and `employee_id` are not needed by any of the code being tested.

you're _probably_ right that your proposed test data more closely matches the intended use-case, but i do not know whether the test data should be minimally viable or whether it should match expected circumstances. i err towards the first.

carmenbianca replied:
on second reading, i agree with you. `hr_timesheet` uses project ids and employees. i will change this.

huguesdk reviewed (changes requested):
just one small change request, otherwise lgtm (only the new module part, the rest is reviewed in #692).

are you planning to rebase this on the #692 branch?

carmenbianca committed:
[FIX] hr_timesheet_begin_end: Test uses timesheet lines now

Timesheet lines are characterised by having a project.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] Add hr_timesheet_begin_end to rebel modules

This is a tricky one. _Technically_ there is nothing that makes
hr_timesheet_begin_end incompatible with the other modules here.
However, when the project_id on a timesheet line is changed, unit_amount
is rerecomputed. This has some consequences.

In the sale_timesheet tests, the project_id of a timesheet line
is (sometimes?) recomputed. Subsequently, unit_amount is recomputed and
set to 0 when it _shouldn't_ be. Under a normal flow this wouldn't be a
problem; time_start and time_stop would be populated to correctly
recompute unit_amount. But in the tests, they are not.

One solution is to, in addition to not recomputing the unit_amount of
non-timesheet lines, not recompute the unit_amount of timesheet lines
whose time_start and time_stop are both set to 00:00. However, this
means that resetting these values to 00:00 does not correctly set
unit_amount back to 0, which seems erroneous to me.

Marking this module as a rebel module seems like the best course of
action to me, even though it would be preferable to test this in
conjunction with the other modules.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] hr_timesheet_begin_end_no_end: Added as rebel module

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk commented on hr_timesheet_begin_end_no_end/__manifest__.py:
```diff
@@ -0,0 +1,22 @@
+
+{
+    "name": "Timesheet - Begin/End Hours with empty end",
+    "summary": """
+        Allow to insert begin hour without end hour, which the user may fill
+        in later.""",
+    "version": "16.0.1.0.0",
+    "category": "Human Resources",
+    "website": "https://github.com/OCA/timesheet",
+    "author": "Coop IT Easy SC, Odoo Community Association (OCA)",
+    "maintainers": ["carmenbianca"],
+    "license": "AGPL-3",
+    "application": False,
+    "depends": ["hr_timesheet_begin_end"],
+    "excludes": [],
+    "data": [],
+    "demo": [],
+    "qweb": [],
```
please remove empty and default values and remove the period at the end of the description. 😉

```suggestion
    "summary": (
        "Allow to insert begin hour without end hour, which the user may "
        "fill in later"
    ),
    "version": "16.0.1.0.0",
    "category": "Human Resources",
    "website": "https://github.com/OCA/timesheet",
    "author": "Coop IT Easy SC, Odoo Community Association (OCA)",
    "maintainers": ["carmenbianca"],
    "license": "AGPL-3",
    "depends": ["hr_timesheet_begin_end"],
```

huguesdk reviewed (approved):
just a cosmetic remark, otherwise lgtm! very clean solution!

huguesdk commented:
@carmenbianca is this ready to be merged?
